src/test/java/io/personium/core/rs/cell/TokenEndPointResourceTest.java

```diff
@@ -516,9 +532,103 @@ public void token_GrantCode_WithConfidentialMark_When_Valid_Succeeds() throws Ex
+        String issuerCellUrl = this.mockCell.getUnitUrl() + "issuerCell/";
+        roleList = new ArrayList<Role>();
+        Role role1 = new Role("MyBoardEditor", "mb", this.mockCell.getUnitUrl() + "appcell/", issuerCellUrl);
+        Role role2 = new Role("MyBoardViewer", "mb", this.mockCell.getUnitUrl() + "appcell/", issuerCellUrl);
+        roleList.add(role1);
+        roleList.add(role2);
+        TransCellAccessToken transCellAccessToken = new TransCellAccessToken(issuerCellUrl, issuerCellUrl + "#me",
+                this.mockCell.getUrl(), roleList, "", new String[0]);
+        log.info(transCellAccessToken.toSamlString());
+
+        // Prepare form contents
+        MultivaluedMap<String, String> formParams = new MultivaluedHashMap<String, String>();
+        formParams.add("grant_type", OAuth2Helper.GrantType.SAML2_BEARER);
+        formParams.add("assertion", transCellAccessToken.toTokenString());
+        formParams.add("client_id", clientId);
+        formParams.add("client_secret", appAuthToken.toTokenString());
+        formParams.add("scope", "root");
+
+        Response res = tokenEndPointResource.token(this.mockUriInfo, null, formParams, xForwadedFor);
+        assertEquals(200, res.getStatus());
+
+        JsonObject json = Json.createReader(new ByteArrayInputStream(((String) res.getEntity()).getBytes()))
+                .readObject();
+        String atStr = json.getString("refresh_token");
+        VisitorRefreshToken vrt = (VisitorRefreshToken) AbstractOAuth2Token.parse(atStr, this.mockCell.getUrl(),
+                this.mockCell.getUnitUrl());
+        for (Role role : vrt.getRoleList()) {
+            log.info(role.toRoleClassURL());
+        }
+        assertEquals(role1.toRoleClassURL(),
+                vrt.getRoleList().get(0).toRoleClassURL());
+        assertEquals(role2.toRoleClassURL(),
+                vrt.getRoleList().get(1).toRoleClassURL());
+
+
+        // Prepare form contents
+        formParams = new MultivaluedHashMap<String, String>();
+        formParams.add("grant_type", OAuth2Helper.GrantType.REFRESH_TOKEN);
+        formParams.add("refresh_token", vrt.toTokenString());
+        formParams.add("client_id", clientId);
+        formParams.add("client_secret", appAuthToken.toTokenString());
+        formParams.add("scope", "root");
+
+        res = tokenEndPointResource.token(this.mockUriInfo, null, formParams, xForwadedFor);
+        assertEquals(200, res.getStatus());
+
+        String tokenStr = json.getString("refresh_token");
+        vrt = (VisitorRefreshToken) AbstractOAuth2Token.parse(tokenStr, this.mockCell.getUrl(),
+                this.mockCell.getUnitUrl());
+        for (Role role : vrt.getRoleList()) {
+            log.info(role.toRoleClassURL());
+        }
+        assertEquals(role1.toRoleClassURL(),
+                vrt.getRoleList().get(0).toRoleClassURL());
+        assertEquals(role2.toRoleClassURL(),
+                vrt.getRoleList().get(1).toRoleClassURL());
```

Comment:
To assert VisitorAccessTokken is necessary.

Reply:
Fixed